File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <vector>
using namespace std;
#include <stdio.h>
#include <iostream>
int solution(vector<int> stones, int k) {
    int answer = 0;
    long long int sn,en,cnt,flag,ans,mid;
    sn = *min_element(stones.begin(),stones.end());
    en = *max_element(stones.begin(),stones.end());
    ans = 0;
    int n = stones.size();
    while (sn <= en){
        mid = sn + (en - sn) / 2;
        cnt = 0;
        flag = 0;
        for (int i = 0; i < n; i++){
            if (stones[i] - (mid - 1) <= 0 ){
                cnt += 1;
            }
            else{
                cnt = 0;
            }
            if (cnt >= k){
                flag = 1;
                break;
                
            }
            
        }
        if (flag){
            en = mid - 1;
        }
        else{
            ans = mid;
            sn = mid + 1;
        }
    }
    return ans;
}

File: 프로그래머스/lv3/64062. 징검다리 건너기/징검다리 건너기.py
def solution(stones, k):
    ans = 0
    n = len(stones)
    sn ,en = min(stones),max(stones)
    # sn 이상 en 이하 까지 검사
    while sn <= en :
        mid = (sn + en) // 2
        # mid 명까지 지나갈수있다 , c
        cnt = 0
        flag = False
        for i in range(n):
            # mid 명이 가면 mid - 1 명까지 지나보내면 -> mid - 1 을 해준다음 생각
            if stones[i] - (mid - 1) <= 0:
                cnt += 1
            else:
                cnt = 0
            if cnt >= k :
                flag = True
                break
        if flag:
            en = mid - 1
        else:
            ans = mid
            sn = mid + 1
    return ans